README.rst

```diff
@@ -1,45 +1,185 @@
-               extra='347-348', court='ca4',
-               canonical_reporter='U.S.', lookup_index=0,
-               token_index=5, reporter_found='U.S.')]
+Here's a full-featured example of efficiently extracting citations from an HTML document and annotating them with
+links to documents in a database.
+
+.. comment
 
+    # mock database model to make the rest of the tutorial executable, in theory:
+    class MyCaseModel:
+        frontend_url = '/us/1/2/'
+        @classmethod
+        def get(cls, citation):
+            return cls()
+
+First our imports::
+
+    # imports
+    from eyecite import get_citations, clean_text, resolve_citations, annotate
+    from eyecite.models import FullCaseCitation
+    from eyecite.resolve import resolve_full_citation
+    from eyecite.tokenizers import HyperscanTokenizer
 
-Options
+We want to insert links into a piece of HTML like this::
+
+    text = """
+        <p>1 <i>U.S.</i> 2, 1 S.Ct. 2.<p>
+        <p>Id.</p>
+        <p>Mass. Gen.    Laws ch. 1, § 2.</p>
+    """
+
+Note that tags may overlap with
+citations and whitespace may be uneven. We want "1 U.S. 2", "1 S.Ct. 2" and "Id." to all
+link to the same URL fetched from our database, since they all refer to the same case.
+Any cites we don't have in our database will link to "/unknown_cite"
+
+First we'll get the text ready for cite extraction::
+
+    cleaned_text = clean_text(text, ['html', 'all_whitespace'])
+
+    # cleaned_text:
+    # "1 U.S. 2, 1 S.Ct. 2. Id. Mass. Gen. Laws ch. 1, § 2."
+
+Next we'll extract citations using a custom tokenizer. Unlike the default
+tokenizer this uses hyperscan for much faster extraction, with a precompiled
+regular expression database stored in ``.test_cache/``.
+This depends on "pip install hyperscan"::
+
+    tokenizer = HyperscanTokenizer(cache_dir=".test_cache")
+    citations = get_citations(cleaned_text, tokenizer=tokenizer)
+
+    # citations:
+    # [
+    #   FullCaseCitation('1 U.S. 2'),
+    #   FullCaseCitation('1 S.Ct. 2'),
+    #   IdCitation(),
+    #   FullLawCitation('Mass. Gen. Laws ch. 1, § 2'),
+    # ]
+
+Now we want to resolve all of the extracted cites into clusters indexed by
+the resource they refer to, such as a case or statute. We'll use a custom
+function to resolve a given full cite to its resource, so we can return our
+own MyCaseModel for citations we recognize. We'll fall back on returning
+:code:`resolve_full_citation()` for citations we don't recognize::
+
+    def resolve_cite(cite):
+        if isinstance(cite, FullCaseCitation):
+            resource = MyCaseModel.get(citation=cite.corrected_citation())
+            if resource:
+                return resource
+        return resolve_full_citation(cite)
+
+    resolutions = resolve_citations(citations, resolve_full_citation=resolve_cite)
+
+    # resolutions:
+    # {
+    #   MyCaseModel('1 U.S. 2'): [FullCaseCitation('1 U.S. 2'), FullCaseCitation('1 S.Ct. 2'), IdCitation()],
```

Comment:
Worth mentioning here what `corrected_citation()` does, and why 1 US 2 == 1 S.Ct 2?